File: DesktopAgent/agent/tools/simulated_annealing_placer.py
"""Simulated annealing-based item placement for UberStrike maps.

The optimiser balances spawn fairness, risk/reward exposure, spacing,
flow adherence, and strategic depth. It operates on a binary walkable mask
plus optional choke/cover metadata and exports placements consumable by Unity.
"""
from __future__ import annotations

import argparse
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional, Sequence, Tuple

import numpy as np

from ..utils.seed import set_global_seed


@dataclass
class PlacementConstraints:
    """Constraints and metadata required for placement optimisation."""

    spawn_points: List[Tuple[float, float]]
    walkable_areas: np.ndarray  # binary mask (H, W)
    choke_points: List[Tuple[float, float]]
    cover_positions: List[Tuple[float, float]]
    existing_items: Dict[str, List[Tuple[float, float]]]


class SimulatedAnnealingPlacer:
    """Optimises item placement using simulated annealing with heuristic scoring."""

    ITEM_RULES: Dict[str, Dict[str, object]] = {
        "weapon_sniper": {"count": 1, "min_spacing": 50.0, "prefer_exposed": True, "respawn_time": 60},
        "weapon_rocket": {"count": 1, "min_spacing": 40.0, "prefer_center": True, "prefer_exposed": True, "respawn_time": 45},
        "weapon_shotgun": {"count": 2, "min_spacing": 30.0, "prefer_enclosed": True, "prefer_cover": True, "respawn_time": 30},
        "armor_heavy": {"count": 1, "min_spacing": 35.0, "prefer_exposed": True, "prefer_center": True, "respawn_time": 45},
        "armor_light": {"count": 3, "min_spacing": 20.0, "prefer_cover": False, "respawn_time": 20},
        "health_mega": {"count": 1, "min_spacing": 40.0, "prefer_low_ground": True, "respawn_time": 35},
        "health_small": {"count": 6, "min_spacing": 15.0, "prefer_paths": True, "respawn_time": 15},
        "ammo_rockets": {"count": 4, "min_spacing": 10.0, "near_weapon": "weapon_rocket", "respawn_time": 20},
        "ammo_bullets": {"count": 6, "min_spacing": 8.0, "scatter": True, "respawn_time": 15},
    }

    def __init__(self, temperature: float = 1000.0, cooling_rate: float = 0.95, seed: Optional[int] = None):
        set_global_seed(seed)
        self.initial_temp = temperature
        self.cooling_rate = cooling_rate
        self.iteration_count = 0
        self.score_history: List[float] = []
        self.rng = np.random.default_rng(seed)

    def optimise(
        self,
        constraints: PlacementConstraints,
        items_to_place: Optional[Dict[str, int]] = None,
        max_iterations: int = 7500,
    ) -> Dict[str, List[Tuple[float, float]]]:
        """Run the annealer; returns a mapping item_type -> positions."""

        if items_to_place is None:
            items_to_place = {k: int(v.get("count", 1)) for k, v in self.ITEM_RULES.items()}

        if constraints.walkable_areas is None or not np.any(constraints.walkable_areas):
            raise ValueError("No walkable cells available for placement.")

        current = self._initialise(constraints, items_to_place)
        current_score = self._evaluate(current, constraints)

        best = {k: list(v) for k, v in current.items()}
        best_score = current_score

        temp = self.initial_temp
        for iteration in range(max_iterations):
            self.iteration_count = iteration
            neighbor = self._neighbor(current, constraints)
            neighbor_score = self._evaluate(neighbor, constraints)

            delta = neighbor_score - current_score
            accept = delta < 0 or self.rng.random() < np.exp(-delta / max(temp, 1e-3))
            if accept:
                current = neighbor
                current_score = neighbor_score
                if current_score < best_score:
                    best = {k: list(v) for k, v in current.items()}
                    best_score = current_score

            temp *= self.cooling_rate
            self.score_history.append(current_score)
            if temp < 0.05 and iteration > 500:
                break

        return best

    # ---- internal helpers -------------------------------------------------

    def _initialise(self, constraints: PlacementConstraints, items: Dict[str, int]) -> Dict[str, List[Tuple[float, float]]]:
        placement: Dict[str, List[Tuple[float, float]]] = {}
        walkable_points = self._walkable_points(constraints.walkable_areas)
        used: List[Tuple[float, float]] = []

        for item_type, count in items.items():
            rule = self.ITEM_RULES.get(item_type, {})
            min_spacing = float(rule.get("min_spacing", 10.0))
            positions: List[Tuple[float, float]] = []
            attempts = 0
            while len(positions) < count and attempts < 2000 and walkable_points:
                idx = int(self.rng.integers(0, len(walkable_points)))
                candidate = tuple(float(c) for c in walkable_points[idx])
                if self._ok_spacing(candidate, used, min_spacing):
                    positions.append(candidate)
                    used.append(candidate)
                attempts += 1
            placement[item_type] = positions
        return placement

    def _neighbor(self, placement: Dict[str, List[Tuple[float, float]]], constraints: PlacementConstraints) -> Dict[str, List[Tuple[float, float]]]:
        neighbor = {k: list(v) for k, v in placement.items()}
        keys = [k for k, v in neighbor.items() if v]
        if not keys:
            return neighbor
        item_type = self.rng.choice(keys)
        item_positions = neighbor[item_type]
        idx = int(self.rng.integers(0, len(item_positions)))
        current_pos = item_positions[idx]

        candidates = self._walkable_points(constraints.walkable_areas)
        if len(candidates) == 0:
            return neighbor

        rule = self.ITEM_RULES.get(item_type, {})
        min_spacing = float(rule.get("min_spacing", 10.0))

        # limit candidate search nearby for local refinement
        nearby = [c for c in candidates if 0 < np.linalg.norm(np.array(c) - np.array(current_pos)) < 25.0]
        if not nearby:
            nearby = candidates
        new_pos = tuple(float(v) for v in nearby[int(self.rng.integers(0, len(nearby)))])

        # spacing check against everything except the moved item
        all_other = [p for k, vals in neighbor.items() for p in vals if not (k == item_type and p == current_pos)]
        if self._ok_spacing(new_pos, all_other, min_spacing):
            neighbor[item_type][idx] = new_pos
        return neighbor

    def _evaluate(self, placement: Dict[str, List[Tuple[float, float]]], constraints: PlacementConstraints) -> float:
        score = 0.0
        score += self._spawn_balance(placement, constraints) * 10.0
        score += self._risk_reward(placement, constraints) * 5.0
        score += self._flow_alignment(placement, constraints) * 3.0
        score += self._spacing_penalty(placement) * 7.0
        score += self._strategic_depth(placement, constraints) * 4.0
        return score

    def _spawn_balance(self, placement: Dict[str, List[Tuple[float, float]]], constraints: PlacementConstraints) -> float:
        if not constraints.spawn_points:
            return 0.0
        advantages: List[float] = []
        for spawn in constraints.spawn_points:
            total = 0.0
            for item, positions in placement.items():
                if not positions:
                    continue
                rule = self.ITEM_RULES.get(item, {})
                value = float(rule.get("respawn_time", 30.0)) / 10.0
                dists = [np.linalg.norm(np.array(spawn) - np.array(p)) for p in positions]
                min_dist = min(dists) if dists else 100.0
                total += value / (min_dist + 1.0)
            advantages.append(total)
        return float(np.std(advantages)) if advantages else 0.0

    def _risk_reward(self, placement: Dict[str, List[Tuple[float, float]]], constraints: PlacementConstraints) -> float:
        score = 0.0
        for item, positions in placement.items():
            rule = self.ITEM_RULES.get(item, {})
            if not positions:
                continue
            if rule.get("prefer_exposed"):
                for pos in positions:
                    if constraints.cover_positions:
                        cover_dists = [np.linalg.norm(np.array(pos) - np.array(c)) for c in constraints.cover_positions]
                        if cover_dists:
                            min_cover = min(cover_dists)
                            if min_cover < 10.0:
                                score += (10.0 - min_cover) * 2.0
            if rule.get("prefer_cover"):
                for pos in positions:
                    if constraints.cover_positions:
                        cover_dists = [np.linalg.norm(np.array(pos) - np.array(c)) for c in constraints.cover_positions]
                        if cover_dists:
                            min_cover = min(cover_dists)
                            if min_cover > 15.0:
                                score += (min_cover - 15.0)
        return score

    def _flow_alignment(self, placement: Dict[str, List[Tuple[float, float]]], constraints: PlacementConstraints) -> float:
        if not constraints.choke_points:
            return 0.0
        score = 0.0
        for positions in placement.values():
            for pos in positions:
                choke_dists = [np.linalg.norm(np.array(pos) - np.array(c)) for c in constraints.choke_points]
                if not choke_dists:
                    continue
                min_choke = min(choke_dists)
                if min_choke < 5.0:
                    score += (5.0 - min_choke) * 3.0
                elif min_choke > 15.0:
                    score += (min_choke - 15.0) * 0.5
        return score

    def _spacing_penalty(self, placement: Dict[str, List[Tuple[float, float]]]) -> float:
        score = 0.0
        all_positions: List[Tuple[str, Tuple[float, float]]] = []
        for item, positions in placement.items():
            rule = self.ITEM_RULES.get(item, {})
            min_spacing = float(rule.get("min_spacing", 10.0))
            for pos in positions:
                for other_item, other_pos in all_positions:
                    dist = np.linalg.norm(np.array(pos) - np.array(other_pos))
                    if dist < min_spacing:
                        score += (min_spacing - dist) * 5.0
                all_positions.append((item, pos))
        return score

    def _strategic_depth(self, placement: Dict[str, List[Tuple[float, float]]], constraints: PlacementConstraints) -> float:
        power_items = []
        for key in ("weapon_sniper", "weapon_rocket", "armor_heavy"):
            power_items.extend(placement.get(key, []))
        if not power_items:
            return 0.0
        center = np.mean(np.array(power_items), axis=0)
        h, w = constraints.walkable_areas.shape
        map_center = np.array([w * 0.5, h * 0.5])
        offset = np.linalg.norm(center - map_center)
        return max(0.0, offset - 30.0)

    # utility helpers -------------------------------------------------------

    @staticmethod
    def _walkable_points(mask: np.ndarray, sample_rate: int = 2) -> List[Tuple[float, float]]:
        ys, xs = np.nonzero(mask)
        points = list(zip(xs.astype(float), ys.astype(float)))
        if sample_rate > 1:
            points = points[::sample_rate]
        return points

    @staticmethod
    def _ok_spacing(point: Tuple[float, float], others: Sequence[Tuple[float, float]], min_spacing: float) -> bool:
        if not others:
            return True
        p = np.array(point)
        return all(np.linalg.norm(p - np.array(o)) >= min_spacing for o in others)

    # exports ---------------------------------------------------------------

    def export_to_json(self, placement: Dict[str, List[Tuple[float, float]]], output_path: str) -> None:
        data = {
            "version": "1.0",
            "algorithm": "simulated_annealing",
            "temperature": self.initial_temp,
            "cooling_rate": self.cooling_rate,
            "iterations": self.iteration_count,
            "final_score": self.score_history[-1] if self.score_history else 0,
            "items": {},
        }
        for item, positions in placement.items():
            data["items"][item] = [{"x": p[0], "y": 0, "z": p[1]} for p in positions]
        Path(output_path).write_text(json.dumps(data, indent=2))


# ------------------------- CLI --------------------------------------------

def _load_walkable_mask(path: Path) -> np.ndarray:
    from PIL import Image

    img = Image.open(path).convert("L")
    arr = np.array(img, dtype=np.uint8)
    return arr > 128


def _load_layout_points(mask: np.ndarray) -> PlacementConstraints:
    h, w = mask.shape
    spawn_points: List[Tuple[float, float]] = [(w * 0.25, h * 0.25), (w * 0.75, h * 0.75)]
    return PlacementConstraints(
        spawn_points=spawn_points,
        walkable_areas=mask,
        choke_points=[(w * 0.5, h * 0.4), (w * 0.5, h * 0.6)],
        cover_positions=[],
        existing_items={},
    )


def main() -> None:
    parser = argparse.ArgumentParser(description="Optimise item placement using simulated annealing")
    parser.add_argument("--map", required=True, help="Path to layout PNG (bright=walkable)")
    parser.add_argument("--output", default="placement.json", help="Output JSON path")
    parser.add_argument("--visualize", action="store_true", help="Show matplotlib plot")
    parser.add_argument("--temp", type=float, default=1000.0, help="Initial temperature")
    parser.add_argument("--cooling", type=float, default=0.95, help="Cooling rate")
    args = parser.parse_args()

    mask = _load_walkable_mask(Path(args.map))
    constraints = _load_layout_points(mask)
    optimiser = SimulatedAnnealingPlacer(args.temp, args.cooling)
    result = optimiser.optimise(constraints)
    optimiser.export_to_json(result, args.output)

    if args.visualize:
        try:
            import matplotlib.pyplot as plt

            plt.imshow(constraints.walkable_areas, cmap="gray", alpha=0.3)
            colors = plt.cm.tab10(np.linspace(0, 1, len(result)))
            for idx, (item, positions) in enumerate(result.items()):
                if positions:
                    xs = [p[0] for p in positions]
                    ys = [p[1] for p in positions]
                    plt.scatter(xs, ys, c=[colors[idx]], s=80, label=item)
            plt.legend()
            plt.title("Simulated Annealing Placement")
            plt.show()
        except Exception:
            print("Visualization skipped (matplotlib not available)")


if __name__ == "__main__":
    main()

File: UberStrike2022/Editor/SimulatedAnnealingPlacer.cs
#if UNITY_EDITOR
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

namespace UnityAI
{
    [Serializable]
    public class PlacementConstraints
    {
        public List<Vector2> spawnPoints = new();
        public List<Vector2> chokePoints = new();
        public List<Vector2> coverPoints = new();
        public bool[,] walkableMask;
        public float cellSize = 1f;
        public Vector3 origin = Vector3.zero;
        public Texture2D heightmap;

        public bool HasWalkable => walkableMask != null && walkableMask.Length > 0;

        public List<Vector3> SampleWalkable(int sampleStride = 2)
        {
            var result = new List<Vector3>();
            if (!HasWalkable)
            {
                return result;
            }

            int width = walkableMask.GetLength(0);
            int height = walkableMask.GetLength(1);
            float halfW = width * cellSize * 0.5f;
            float halfH = height * cellSize * 0.5f;

            for (int y = 0; y < height; y += Mathf.Max(1, sampleStride))
            {
                for (int x = 0; x < width; x += Mathf.Max(1, sampleStride))
                {
                    if (!walkableMask[x, y])
                    {
                        continue;
                    }
                    float wx = x * cellSize - halfW + cellSize * 0.5f;
                    float wz = halfH - y * cellSize - cellSize * 0.5f;
                    float wy = SampleHeight(x, y);
                    result.Add(origin + new Vector3(wx, wy, wz));
                }
            }

            return result;
        }

        public float SampleHeight(int px, int py)
        {
            if (!heightmap)
            {
                return 0f;
            }
            px = Mathf.Clamp(px, 0, heightmap.width - 1);
            py = Mathf.Clamp(py, 0, heightmap.height - 1);
            var c = heightmap.GetPixel(px, py);
            return c.grayscale;
        }
    }

    [Serializable]
    public class ItemPlacementRule
    {
        public string key;
        public GameObject prefab;
        public int count;
        public float minSpacing;
        public bool preferCover;
        public bool preferExposed;
        public bool preferCenter;
    }

    public static class SimulatedAnnealingPlacer
    {
        private const float DEFAULT_TEMP = 750f;
        private const float DEFAULT_COOLING = 0.96f;
        private const int DEFAULT_ITERATIONS = 4500;

        public static Dictionary<string, List<Vector3>> Optimise(
            PlacementConstraints constraints,
            List<ItemPlacementRule> rules,
            int maxIterations = DEFAULT_ITERATIONS,
            float initialTemperature = DEFAULT_TEMP,
            float coolingRate = DEFAULT_COOLING)
        {
            var result = new Dictionary<string, List<Vector3>>();
            if (constraints == null || rules == null || rules.Count == 0 || !constraints.HasWalkable)
            {
                return result;
            }

            var rand = new System.Random(1337);
            var walkable = constraints.SampleWalkable(2);
            if (walkable.Count == 0)
            {
                return result;
            }

            var current = Initialise(rules, walkable, rand);
            float currentScore = Evaluate(current, constraints, rules);
            var best = Clone(current);
            float bestScore = currentScore;

            float temp = initialTemperature;
            for (int iteration = 0; iteration < maxIterations; iteration++)
            {
                var neighbor = Neighbor(current, constraints, walkable, rules, rand);
                float neighborScore = Evaluate(neighbor, constraints, rules);
                float delta = neighborScore - currentScore;
                bool accept = delta < 0f || UnityEngine.Random.value < Mathf.Exp(-delta / Mathf.Max(0.01f, temp));
                if (accept)
                {
                    current = neighbor;
                    currentScore = neighborScore;
                    if (neighborScore < bestScore)
                    {
                        best = Clone(neighbor);
                        bestScore = neighborScore;
                    }
                }

                temp *= coolingRate;
                if (temp < 0.05f && iteration > 800)
                {
                    break;
                }
            }

            return best;
        }

        private static Dictionary<string, List<Vector3>> Initialise(List<ItemPlacementRule> rules, List<Vector3> candidates, System.Random rand)
        {
            var placement = new Dictionary<string, List<Vector3>>();
            var used = new List<Vector3>();
            foreach (var rule in rules)
            {
                var list = new List<Vector3>();
                int attempts = 0;
                while (list.Count < rule.count && attempts < 1200 && candidates.Count > 0)
                {
                    var p = candidates[rand.Next(candidates.Count)];
                    if (OkSpacing(p, used, rule.minSpacing))
                    {
                        list.Add(p);
                        used.Add(p);
                    }
                    attempts++;
                }
                placement[rule.key] = list;
            }
            return placement;
        }

        private static Dictionary<string, List<Vector3>> Neighbor(
            Dictionary<string, List<Vector3>> current,
            PlacementConstraints constraints,
            List<Vector3> candidates,
            List<ItemPlacementRule> rules,
            System.Random rand)
        {
            var result = Clone(current);
            var movableKeys = result.Where(kvp => kvp.Value.Count > 0).Select(kvp => kvp.Key).ToList();
            if (movableKeys.Count == 0)
            {
                return result;
            }

            string key = movableKeys[rand.Next(movableKeys.Count)];
            var positions = result[key];
            int idx = rand.Next(positions.Count);
            var currentPos = positions[idx];
            var rule = rules.FirstOrDefault(r => r.key == key);
            float minSpacing = rule != null ? rule.minSpacing : 10f;

            var nearby = candidates.Where(c => Vector3.Distance(c, currentPos) < 20f && Vector3.Distance(c, currentPos) > 0.1f).ToList();
            if (nearby.Count == 0)
            {
                nearby = candidates;
            }

            var candidate = nearby[rand.Next(nearby.Count)];
            var other = result.Where(kvp => kvp.Key != key).SelectMany(kvp => kvp.Value).ToList();
            other.AddRange(positions.Where((p, i) => i != idx));
            if (OkSpacing(candidate, other, minSpacing))
            {
                positions[idx] = candidate;
            }
            return result;
        }

        private static float Evaluate(Dictionary<string, List<Vector3>> placement, PlacementConstraints constraints, List<ItemPlacementRule> rules)
        {
            float score = 0f;
            score += SpawnBalance(placement, constraints, rules) * 10f;
            score += RiskReward(placement, constraints, rules) * 5f;
            score += FlowAlignment(placement, constraints) * 3f;
            score += SpacingPenalty(placement, rules) * 7f;
            score += StrategicDepth(placement, constraints) * 4f;
            return score;
        }

        private static float SpawnBalance(Dictionary<string, List<Vector3>> placement, PlacementConstraints constraints, List<ItemPlacementRule> rules)
        {
            if (constraints.spawnPoints == null || constraints.spawnPoints.Count == 0)
            {
                return 0f;
            }

            var advantages = new List<float>();
            foreach (var spawn in constraints.spawnPoints)
            {
                float total = 0f;
                foreach (var kvp in placement)
                {
                    var rule = rules.FirstOrDefault(r => r.key == kvp.Key);
                    float value = rule != null ? Mathf.Max(1f, rule.count) : 1f;
                    if (kvp.Value.Count == 0)
                    {
                        continue;
                    }
                    float minDist = kvp.Value.Min(p => Vector2.Distance(new Vector2(p.x, p.z), spawn));
                    total += value / (minDist + 1f);
                }
                advantages.Add(total);
            }
            return Std(advantages);
        }

        private static float RiskReward(Dictionary<string, List<Vector3>> placement, PlacementConstraints constraints, List<ItemPlacementRule> rules)
        {
            float score = 0f;
            foreach (var kvp in placement)
            {
                var rule = rules.FirstOrDefault(r => r.key == kvp.Key);
                if (rule == null)
                {
                    continue;
                }

                foreach (var pos in kvp.Value)
                {
                    if (rule.preferExposed && constraints.coverPoints.Count > 0)
                    {
                        float cover = constraints.coverPoints.Min(c => Vector2.Distance(c, new Vector2(pos.x, pos.z)));
                        if (cover < 8f)
                        {
                            score += (8f - cover) * 2f;
                        }
                    }
                    if (rule.preferCover && constraints.coverPoints.Count > 0)
                    {
                        float cover = constraints.coverPoints.Min(c => Vector2.Distance(c, new Vector2(pos.x, pos.z)));
                        if (cover > 14f)
                        {
                            score += (cover - 14f);
                        }
                    }

                    if (rule.preferCenter && constraints.walkableMask != null)
                    {
                        float cx = constraints.walkableMask.GetLength(0) * constraints.cellSize * 0.5f;
                        float cz = constraints.walkableMask.GetLength(1) * constraints.cellSize * 0.5f;
                        float distCenter = Vector2.Distance(new Vector2(pos.x, pos.z), new Vector2(cx, cz));
                        if (distCenter > 25f)
                        {
                            score += (distCenter - 25f) * 0.5f;
                        }
                    }
                }
            }
            return score;
        }

        private static float FlowAlignment(Dictionary<string, List<Vector3>> placement, PlacementConstraints constraints)
        {
            if (constraints.chokePoints == null || constraints.chokePoints.Count == 0)
            {
                return 0f;
            }
            float score = 0f;
            foreach (var kvp in placement)
            {
                foreach (var pos in kvp.Value)
                {
                    float minChoke = constraints.chokePoints.Min(c => Vector2.Distance(c, new Vector2(pos.x, pos.z)));
                    if (minChoke < 5f)
                    {
                        score += (5f - minChoke) * 3f;
                    }
                    else if (minChoke > 15f)
                    {
                        score += (minChoke - 15f) * 0.5f;
                    }
                }
            }
            return score;
        }

        private static float SpacingPenalty(Dictionary<string, List<Vector3>> placement, List<ItemPlacementRule> rules)
        {
            float score = 0f;
            var all = new List<(Vector3 pos, float spacing)>();
            foreach (var kvp in placement)
            {
                var rule = rules.FirstOrDefault(r => r.key == kvp.Key);
                float minSpacing = rule != null ? rule.minSpacing : 10f;
                foreach (var pos in kvp.Value)
                {
                    foreach (var other in all)
                    {
                        float dist = Vector3.Distance(pos, other.pos);
                        float req = Mathf.Max(minSpacing, other.spacing);
                        if (dist < req)
                        {
                            score += (req - dist) * 5f;
                        }
                    }
                    all.Add((pos, minSpacing));
                }
            }
            return score;
        }

        private static float StrategicDepth(Dictionary<string, List<Vector3>> placement, PlacementConstraints constraints)
        {
            var power = new List<Vector2>();
            foreach (var key in new[] { "weapon_sniper", "weapon_rocket", "armor_heavy" })
            {
                if (placement.TryGetValue(key, out var list))
                {
                    power.AddRange(list.Select(p => new Vector2(p.x, p.z)));
                }
            }
            if (power.Count == 0)
            {
                return 0f;
            }

            float cx = constraints.walkableMask.GetLength(0) * constraints.cellSize * 0.5f;
            float cz = constraints.walkableMask.GetLength(1) * constraints.cellSize * 0.5f;
            var center = new Vector2(cx, cz);
            var bary = Vector2.zero;
            foreach (var p in power)
            {
                bary += p;
            }
            bary /= power.Count;
            float offset = Vector2.Distance(center, bary);
            return Mathf.Max(0f, offset - 30f);
        }

        private static bool OkSpacing(Vector3 point, List<Vector3> others, float minSpacing)
        {
            foreach (var o in others)
            {
                if (Vector3.Distance(point, o) < minSpacing)
                {
                    return false;
                }
            }
            return true;
        }

        private static float Std(List<float> values)
        {
            if (values == null || values.Count == 0)
            {
                return 0f;
            }
            float mean = values.Sum() / values.Count;
            float variance = values.Sum(v => (v - mean) * (v - mean)) / values.Count;
            return Mathf.Sqrt(variance);
        }

        private static Dictionary<string, List<Vector3>> Clone(Dictionary<string, List<Vector3>> src)
        {
            var dst = new Dictionary<string, List<Vector3>>();
            foreach (var kvp in src)
            {
                dst[kvp.Key] = new List<Vector3>(kvp.Value);
            }
            return dst;
        }
    }
}
#endif
